Allow signs to update price from worth

moxventura opened the pull request:
closes #2638

Commit:
Allow signs to update price from worth

closes #3332

Commit:
Add translations for price changes on signs

JRoy commented on Essentials/src/com/earth2me/essentials/Settings.java:
```diff
@@ -713,6 +716,41 @@ public boolean isEcoDisabled() {
         return economyDisabled;
     }
 
+    private boolean buySignUpdatedToWorth = false;
+
+    private boolean _getBuySignUpdatedToWorth() {
+      return config.getBoolean("economy-update-buy-sign-to-worth", false);
+    }
+
+    @Override
+    public boolean isBuySignUpdatedToWorth() {
+      return buySignUpdatedToWorth;
+    }
+
+    private boolean sellSignUpdatedToWorth = false;
+
+    private boolean _getSellSignUpdatedToWorth() {
+      return config.getBoolean("economy-update-sell-sign-to-worth", false);
+    }
+
+    @Override
+    public boolean isSellSignUpdatedToWorth() {
+      return sellSignUpdatedToWorth;
+    }
```
These config options do not require a helper method, you can inline the config.getBoolean call
```suggestion
    @Override
    public boolean isBuySignUpdatedToWorth() {
      return config.getBoolean("economy-update-buy-sign-to-worth", false);
    }

    @Override
    public boolean isSellSignUpdatedToWorth() {
      return config.getBoolean("economy-update-sell-sign-to-worth", false);
    }
```

moxventura replied:
I was wondering where those helper functions were for :)

JRoy commented on Essentials/src/com/earth2me/essentials/Settings.java:
```diff
@@ -547,6 +547,9 @@ public void reloadConfig() {
         economyLog = _isEcoLogEnabled();
         economyLogUpdate = _isEcoLogUpdateEnabled();
         economyDisabled = _isEcoDisabled();
+        sellSignUpdatedToWorth = _getSellSignUpdatedToWorth();
+        buySignUpdatedToWorth = _getBuySignUpdatedToWorth();
```
Reflects changes below
```suggestion

```

JRoy commented on Essentials/src/com/earth2me/essentials/Settings.java:
```diff
@@ -713,6 +716,41 @@ public boolean isEcoDisabled() {
+      return sellSignUpdatedToWorth;
+    }
+
+    private BigDecimal economyBuyMultiplier = new BigDecimal("1.0");
```
Don't initialize variables here
```suggestion
    private BigDecimal economyBuyMultiplier;
```

JRoy commented on Essentials/src/messages_en.properties:
```diff
@@ -421,6 +421,7 @@ powerToolRemove=§6Command §c{0}§6 removed from §c{1}§6.
 powerToolRemoveAll=§6All commands removed from §c{0}§6.
 powerToolsDisabled=§6All of your power tools have been disabled.
 powerToolsEnabled=§6All of your power tools have been enabled.
+priceChanged=§6Price changed from {0} to {1}.
```
Please do not add this here, Crowdin will automatically do this
```suggestion
```

moxventura replied:
didn't know that

JRoy commented on Essentials/src/messages_nl.properties:
```diff
@@ -421,6 +421,7 @@ powerToolRemove=§6Commando §c{0}§6 verwijderd van §c{1}§6.
 powerToolRemoveAll=§6Alle commandos zijn verwijderd van §c{0}§6.
 powerToolsDisabled=§6Al uw powertools zijn uitgeschakeld.
 powerToolsEnabled=§6Al uw powertools zijn ingeschakeld.
+priceChanged=§6Prijs is veranderd van {0} naar {1}.
```
Please do not add this here, Crowdin will automatically do this
```suggestion
```

JRoy commented on Essentials/src/messages.properties:
```diff
@@ -427,6 +427,7 @@ powerToolRemove=\u00a76Command \u00a7c{0}\u00a76 removed from \u00a7c{1}\u00a76.
 powerToolRemoveAll=\u00a76All commands removed from \u00a7c{0}\u00a76.
 powerToolsDisabled=\u00a76All of your power tools have been disabled.
 powerToolsEnabled=\u00a76All of your power tools have been enabled.
+priceChanged=\u00a76Price changed from {0} to {1}.
```
I feel like this message could be a little more useful/explanatory

moxventura replied:
I could add the entity name

JRoy commented on Essentials/src/config.yml:
```diff
@@ -692,6 +692,16 @@ max-money: 10000000000000
+# Enable this to make buy signs automatically adjust their price according to worth.yml
+economy-update-buy-sign-to-worth: false
+
+# If you don't want buy and sell prices to be the same, set this multiplier to something higher then 1.0
+# All buy prices for players will be multiplied by this amount.
```
This description should be improved. It's not clear if this only happens when the above booleans are enabled.

moxventura replied:
true, I'll fix it. This multiplier could also be used if one would implement the 'buy' method (like in #2795). But for now I'll add more information to it.

JRoy commented on Essentials/src/config.yml:
```diff
@@ -692,6 +692,16 @@ max-money: 10000000000000
 # Setting this to 0, will disable overdrafts/loans completely.  Users need 'essentials.eco.loan' perm to go below 0.
 min-money: -10000
 
+# Enable this to make sell signs automatically adjust their price according to worth.yml
+economy-update-sell-sign-to-worth: false
+
+# Enable this to make buy signs automatically adjust their price according to worth.yml
+economy-update-buy-sign-to-worth: false
```
Not a fan of the descriptions here, 'automatically' implies that all signs would update upon reloading

moxventura replied:
I'll change this

JRoy commented on Essentials/src/com/earth2me/essentials/signs/SignSell.java:
```diff
@@ -18,13 +18,24 @@ public SignSell() {
-        validateTrade(sign, 1, 2, player, ess);
+        validateTrade(sign, 1, 2, player, ess);  
+        if(ess.getSettings().isSellSignUpdatedToWorth()) {
+            try {
+                updateFromWorth(sign, ess);
+            }
+            catch (SignException e) {
+                // It's fine
+            }
```
No, it's not fine?

JRoy commented on Essentials/src/com/earth2me/essentials/signs/SignBuy.java:
```diff
@@ -18,12 +17,23 @@ public SignBuy() {
     @Override
     protected boolean onSignCreate(final ISign sign, final User player, final String username, final IEssentials ess) throws SignException {
         validateTrade(sign, 1, 2, player, ess);
+        if(ess.getSettings().isBuySignUpdatedToWorth()) {
+            try {
+              updateFromWorth(sign, ess, ess.getSettings().getEcoBuyMultiplier());
+            }
+            catch (SignException e) {
+              // It's fine
+            }
+        }
```
No, it's not fine?

moxventura replied:
In this location this means that the price could not be found in worth.yml If we do nothing here the sign will work as it used to. Should I elaborate on that in the comment?

JRoy replied:
There should be a check in the method to validate this, not a try catch here

moxventura replied:
Could you help me out how I should do that? 
The solution in my head now is to pass a boolean whether or not an exception should be thrown. But I'm not fond if this idea. I'll have to know the difference between a 'create' update and a 'use' update if I want to be able to make the decision whether or not to throw the exception.

Technoguyfication replied:
In general, it's bad practice to throw exceptions during normal operation of a program. They're computationally expensive and make the code harder to understand when an exception is "supposed" to be thrown. You should return a value like a boolean that indicates whether the price changed, and then if the price _did_ change, cancel the buy/sell and let the player know about the price change so they can decide whether to complete their transaction at the new price.

moxventura replied:
Fixed

JRoy commented on Essentials/src/com/earth2me/essentials/signs/EssentialsSign.java:
```diff
@@ -436,6 +436,30 @@ protected final Trade getTrade(final ISign sign, final int index, final int decr
         }
     }
 
+    /**
+     * Updates the sign cost to the value of the item specified in worth.yml
+     */
+    protected void updateFromWorth(final ISign sign, final IEssentials ess) throws SignException {
+      updateFromWorth(sign, ess, new BigDecimal("1"));
+    }
+    
+    /**
+     * Updates the sign cost to the value of the item specified in worth.yml multiplied by a multiplier factor
+     */
+    protected void updateFromWorth(final ISign sign, final IEssentials ess, final BigDecimal multiplier) throws SignException {
+      final ItemStack stack = getItemStack(getSignText(sign, 2), getIntegerPositive(sign.getLine(1)), ess);
+      final int amount = stack.getAmount();
+      BigDecimal price = ess.getWorth().getPrice(ess, stack);
+      if (price == null || amount == 0) return;
+      price = price.multiply(multiplier).multiply(new BigDecimal(amount));
+      final BigDecimal oldPrice = getMoney(getSignText(sign, 3), ess);
+      if ( oldPrice == null || price.compareTo(oldPrice) != 0) {
+        sign.setLine(3, NumberUtil.shortCurrency(price, ess));
+        sign.updateSign();
+        throw new SignException(tl("priceChanged", oldPrice, price));
+      }
+  }
```
Why the two separate methods here?

moxventura replied:
A habit, I can change it to one...

JRoy commented on Essentials/src/com/earth2me/essentials/signs/EssentialsSign.java:
```diff
@@ -436,6 +436,30 @@ protected final Trade getTrade(final ISign sign, final int index, final int decr
+      final ItemStack stack = getItemStack(getSignText(sign, 2), getIntegerPositive(sign.getLine(1)), ess);
+      final int amount = stack.getAmount();
+      BigDecimal price = ess.getWorth().getPrice(ess, stack);
+      if (price == null || amount == 0) return;
```
```suggestion
      if (price == null || amount == 0) {
        return;
      }
```

JRoy commented on Essentials/src/com/earth2me/essentials/signs/EssentialsSign.java:
```diff
@@ -436,6 +436,30 @@ protected final Trade getTrade(final ISign sign, final int index, final int decr
+      if (price == null || amount == 0) return;
+      price = price.multiply(multiplier).multiply(new BigDecimal(amount));
+      final BigDecimal oldPrice = getMoney(getSignText(sign, 3), ess);
+      if ( oldPrice == null || price.compareTo(oldPrice) != 0) {
```
```suggestion
      if (oldPrice == null || price.compareTo(oldPrice) != 0) {
```

JRoy commented on Essentials/src/com/earth2me/essentials/signs/EssentialsSign.java:
```diff
@@ -436,6 +436,30 @@ protected final Trade getTrade(final ISign sign, final int index, final int decr
+      if ( oldPrice == null || price.compareTo(oldPrice) != 0) {
+        sign.setLine(3, NumberUtil.shortCurrency(price, ess));
+        sign.updateSign();
+        throw new SignException(tl("priceChanged", oldPrice, price));
```
The message will never be sent if you always ignore it.

moxventura replied:
if I test this I see this message. The exception is caught somewhere and shown to the player. I assumed this was how it is supposed to work. Maybe it is a more elegant solution if I catch this somewhere else and display the message there?

moxventura replied:
This is caught on line 64 and 118, I think that's fine

JRoy reviewed (changes requested):
Overall not a fan of this feature but would be happy to hear the input of others @md678685 @pop4959

moxventura commented:
Thanks for the review! I'll fix the issues..

moxventura committed:
Apply suggestions from code review

Co-authored-by: Josh Roy <10731363+JRoy@users.noreply.github.com>

moxventura committed:
Apply suggestions from code review

Co-authored-by: Josh Roy <10731363+JRoy@users.noreply.github.com>

pop4959 reviewed (changes requested):
@JRoy I'm not so fond of this feature either, and frankly not sure why we haven't closed the original issue yet. There are several problems, both logistically with how the signs are meant to work, as well as with implementing it.

First of all, I want to mention that in the original issue I suggested that this would be a tricky feature to support since Essentials would need to keep track of buy/sell signs that are created. This would need to be done so that prices can be updated automatically. I do appreciate the fact that @moxventura attempted to go with a storage-less implementation, however it is very apparent that the signs 1) do not update automatically, and thus 2) will confuse or aggravate players trying to use them, since they will receive an "error" about the price changing right as they try to use it.

Now, to point out a few flaws with the feature request itself. It suggests a use case in which players set up their own buy/sell signs, which to me sounds like complete nonsense. These signs are intended to be created by server admins only, as they contain an infinite stock. Pulling automatically from worth isn't very useful, since usually prices are set higher than their normal sell price (which in this PR's implementation is re-solved by adding another setting for buy multiplier). Obviously the suggestion was to change the behavior so that players can create the signs without being able to set the price themselves (such as a zero amount) and bypassing worth. However, there is already the /sell command which meets this purpose (which the person suggesting the issue was deliberately trying to avoid using for whatever reason). Not to mention that requiring players to set up a sell sign just to sell an item, instead of using the /sell command just seems plain silly...

There is nothing wrong with trade signs, which are meant to be set up and used by players. They can set their own prices, but this is fine because the items they are selling are theirs to sell.

I'd be curious to hear more from @moxventura about why they were interested in making this PR. Personal interest, or just to help tackle some Essentials issues? (which, by the way is very much appreciated!) For future PRs, I'd advise checking in with a team member before making a PR for a larger feature (especially one that isn't tagged with "open to pr"), especially when there have been concerns voiced around it.

Commit:
Apply suggestions from review

Commit:
Elaborate on price change message

moxventura commented:
@pop4959 thanks for the reply! I'd love to elaborate. First of all, I agree completely with you that it's not the goal to have signs that are placed by the players itself. It was suggested in the original post but I agree that the signs should be placed by ops. 
I was triggered to create this PR due to the comment of @Technoguyfication in the issue since I had the same problem myself. 

The goal is to allow easy change of global prices without having moderators change all the signs in your world and find out you forgot one or two which people are abusing to gain infinite money. This will allow moderators to balance the economy during gameplay on a global scale depending on the needs of the server. (Brainfart: ultimately you could let the supply and demand calculate prices and create a real simulation of an economy. Maybe I'll get my hands on this idea later on).

My second goal was to leave the original operation in tact so this is an added feature which is turned off by default and does no harm to people who won't use it.

The way the current PR is implemented will allow server admins to do a single command (/setworth [item] [price]) and the prices will change on all signs. It's true that the first user who tries to use a sign with an old price will get a message of the failed transfer, but it will be clear the price has changed and after that first contact the price on the sign will be correct and the sign will function as expected. But to be honest, this is how an economy works.. prices change.

I'd love to help out with other issues as well, I had a lot of fun figuring out how to create this one and I already use it on my server. Also having my code reviewed is really educational. Thanks!

Techno3600 commented:
Hey, so while some might find this feature useless. I certainly do not. I don’t think many understand the pain of going through an entire worth.yml only to find you have to also hardcode those values into the signs. I haven’t tested this out but these should only work with admin shop signs.

Also, could you make an option to have these only affect the sell line in signs. Currently, my server marks up the buy prices of items so it would be perfect to just use this and then manually mark up the buy price.

Also, I understand the concern that this would be frustrating to players when it updates on the spot, but my management would just go through and hit all the signs in my shop once (because we know what has admin signs from the worth.yml).

I understand why some don’t like this idea. But I think there is a certain convenience factor with this change that it’s almost dumb not to implement it.

I am a programmer but haven’t delved deep into Minecraft plugins yet. Would it be too difficult to find all the signs in the world, check to see if they’re valid admin shop signs, and then update them to the correct prices? If that’s not possible, the current idea you have where it updates on click is fine for me since we’re going to hit it anyway before a change goes live.

Thanks,
Techno

moxventura commented:
Hey, thanks for the feedback. If you are interested I could send you the compiled jar (or you could compile it yourself) to see how it works. I really think players won't mind seeing the price-change message.

> Also, could you make an option to have these only affect the sell line in signs. Currently, my server marks up the buy prices of items so it would be perfect to just use this and then manually mark up the buy price.

I'm not sure what you mean, but if you mean only sell-signs then the feature is already in there. These are the config settings I added:
```
# Enable this to make sell signs adjust their price according to worth.yml when they are created or used
economy-update-sell-sign-to-worth: false

# Enable this to make buy signs adjust their price according to worth.yml when they are created or used
economy-update-buy-sign-to-worth: false

# If you don't want buy and sell prices to be the same, set this multiplier to something higher then 1.0
# All buy prices on buy signs will be the worth value multiplied by this amount.
economy-buy-multiplier: 1.0
```
So you could only set this feature for the sell signs.
I did not touch the trade-signs...

> I am a programmer but haven’t delved deep into Minecraft plugins yet. Would it be too difficult to find all the signs in the world, check to see if they’re valid admin shop signs, and then update them to the correct prices? If that’s not possible, the current idea you have where it updates on click is fine for me since we’re going to hit it anyway before a change goes live.

I did play with this idea for a while but from what I understood is that it would mean to check every block in the server to see if it's a sign. And you'll need to do that every time a price has changed. That's a performance hit, even if you do it on chunk-load. You might be better off doing it offline using a level editor. Another option would be to keep a list of signs somewhere, but that's a lot of work I think.

Technoguyfication commented:
Maybe we could differentiate signs that have a fixed worth and signs that have a dynamic worth from `/setworth` or `worth.yml` by putting the price on the sign in brackets, for example. This would let old signs function as normal and keep their old prices, and the plugin can also check to see if line 4 of the sign is inside brackets before updating the sign price. That way, there's no "all-or-nothing" option in the config to convert _all_ signs in the world, and instead let admins choose which signs update automatically.

Example:
```
[Buy]
64
cobblestone
$64
```
...would act as a normal sign and the price of $64 never changes. For a dynamic sign, it looks like this:
```
[Buy]
64
cobblestone
[$64]
```
This sign will update automatically. If the price of cobblestone is set to $2, the next time the sign is interacted with, it will update to `[$128]`

pop4959 commented:
I don't have time to give a full response currently but I'm glad we're on the same page about buy/sell signs clearly being an admin feature. Also, I really do think the implementation of this could be better. I was planning to suggest something similar to what @Technoguyfication just mentioned, with a sign syntax that indicates the price is set by worth. This would allow an admin to selectively create signs that use worth, while also not limiting them to only using worth.

Techno3600 commented:
I am so happy you also have a buy multiplayer. You have no idea how frustrating it was for our management team to have to make a whole worth.yml and then hardcode all the values in. I think they will be happy to hear a change is in the works!

Techno3600 commented:
@moxventura Also, when a user clicks on the sign when it's not updated to the new price in the worth.yml, will it charge them the new amount and then tell them the price has changed? Or will it cancel the transaction, tell them the price has changed, update the sign, and then allow them to recomplete the transaction?

moxventura commented:
> @moxventura Also, when a user clicks on the sign when it's not updated to the new price in the worth.yml, will it charge them the new amount and then tell them the price has changed? Or will it cancel the transaction, tell them the price has changed, update the sign, and then allow them to recomplete the transaction?

@Techno3600 The latter, let's give them the opportunity to reconsider their purchase. 
[![mc-signs.gif](https://i.postimg.cc/KY2Jr78t/mc-signs.gif)](https://postimg.cc/9DLdc7Tf)

Commit:
Remove try-catch on sign create

Commit:
Round prices to two decimals

Commit:
Change to optional worth signs using colors

moxventura commented:
Ok, so here's what I've come up with. @Technoguyfication @pop4959 

If you leave the 'cost' line empty on placing a sign: it will automatically use the price from worth.yml
Otherwise it will use the cost which is provided.
If the item is not found in worth.yml it will keep the current behaviour (which is buy/sell the items for free, which might be a bug?)

Anyway, if a price from worth.yml is used, the price is colored (yellow). And will update the price as soon as the price has changed. Also see the attached gif:

https://i.postimg.cc/MK14NT6J/mc-signs.gif

pop4959 commented on Essentials/src/config.yml:
```diff
@@ -709,6 +709,10 @@ max-money: 10000000000000
 
+# If you don't want buy and sell prices to be the same, set this multiplier to something higher then 1.0
+# All buy prices on buy signs will be the worth value multiplied by this amount.
+economy-buy-multiplier: 1.0
```
maybe just "buy-multiplier" since this is already in the economy section

moxventura replied:
makes sense

pop4959 commented on Essentials/src/messages.properties:
```diff
@@ -583,6 +583,7 @@ powerToolRemove=\u00a76Command \u00a7c{0}\u00a76 removed from \u00a7c{1}\u00a76.
 powerToolRemoveAll=\u00a76All commands removed from \u00a7c{0}\u00a76.
 powerToolsDisabled=\u00a76All of your power tools have been disabled.
 powerToolsEnabled=\u00a76All of your power tools have been enabled.
+priceChanged=\u00a76The price of \u00a7c{0}x {1}\u00a76 has changed from \u00a7e{2}\u00a76 to \u00a7e{3}\u00a76.
```
should probably not be displayed as an error, since it isn't one. also, normally c is used for variables, not e.

moxventura replied:
changed

pop4959 reviewed (changes requested):
Not a full review, just some more comments

Commit:
Remove chatcolor before calculating sign value

Commit:
Fix changed-message not being an Error

Changed message colors to be compatible
Show prices in nice formatting
Changed economy-buy-multiplier to buy-multiplier

mdcfe commented on Essentials/src/com/earth2me/essentials/signs/EssentialsSign.java:
```diff
@@ -375,7 +376,8 @@ protected final ItemStack getItemMeta(final ItemStack item, final String meta, f
     }
 
-    protected final BigDecimal getMoney(final String line, final IEssentials ess) throws SignException {
+    protected final BigDecimal getMoney(String line, final IEssentials ess) throws SignException {
+        line = ChatColor.stripColor(line);
```
Why is this change here?

moxventura replied:
I color the lines if it's updated from worth. So you can see which signs use the values from worth.yml. If I don't strip the color here we won't be able to get the price value since there could be non-number (color) characters on this line.

mdcfe commented on Essentials/src/com/earth2me/essentials/signs/EssentialsSign.java:
```diff
@@ -435,6 +437,33 @@ protected final Trade getTrade(final ISign sign, final int index, final int decr
             return new Trade(money, ess);
         }
     }
+      
+    /**
+     * Updates the sign cost to the value of the item specified in worth.yml multiplied by a multiplier factor
+     * @param sign
+     * @param player
+     * @param ess
+     * @param multiplier multiply the worth price by this amount
+     * 
+     * @returns true if the price was changed, false otherwise
+     */
+    protected boolean updateFromWorth(final ISign sign, final User player, final IEssentials ess, final BigDecimal multiplier) throws SignException {
+        final ItemStack stack = getItemStack(getSignText(sign, 2), getIntegerPositive(sign.getLine(1)), ess);
+        final int amount = stack.getAmount();
+        BigDecimal price = ess.getWorth().getPrice(ess, stack);
+        if (price != null && amount > 0) {
+            price = price.multiply(multiplier).multiply(new BigDecimal(amount)).setScale(2, RoundingMode.UP);
+            final BigDecimal oldPrice = getMoney(getSignText(sign, 3), ess);
+            if (oldPrice == null || price.compareTo(oldPrice) != 0) {
+                final String priceString = NumberUtil.shortCurrency(price, ess);
+                sign.setLine(3, tl("signFormatWorth") + priceString);
+                sign.updateSign();
+                player.sendMessage(tl("priceChanged", amount, stack.getType().toString().toLowerCase(Locale.ENGLISH), (oldPrice == null ? tl("none") : NumberUtil.shortCurrency(oldPrice, ess)), priceString));
+                return true;
+            }
+        }
+        return false;
+  }
```
This behaviour is specific to buy and sell signs - I'd be inclined to put this in an `EssentialsShopSign` subclass.

moxventura replied:
Done

mdcfe commented on Essentials/src/com/earth2me/essentials/signs/SignBuy.java:
```diff
@@ -18,12 +19,20 @@ public SignBuy() {
     protected boolean onSignCreate(final ISign sign, final User player, final String username, final IEssentials ess) throws SignException {
         validateTrade(sign, 1, 2, player, ess);
-        validateTrade(sign, 3, ess);
+        if(sign.getLine(3).isEmpty()) {
+            updateFromWorth(sign, player, ess, ess.getSettings().getEcoBuyMultiplier());
+        }
+        else {
```
```suggestion
        if (sign.getLine(3).isEmpty()) {
            updateFromWorth(sign, player, ess, ess.getSettings().getEcoBuyMultiplier());
        } else {
```

moxventura replied:
you cuddle the else? It's a matter of taste I guess :)

pop4959 replied:
you mean the standard format that everyone uses? lol

moxventura replied:
Guess I'm old school ¯\_(ツ)_/¯

mdcfe commented on Essentials/src/com/earth2me/essentials/signs/SignBuy.java:
```diff
@@ -18,12 +19,20 @@ public SignBuy() {
 
     @Override
     protected boolean onSignInteract(final ISign sign, final User player, final String username, final IEssentials ess) throws SignException, ChargeException, MaxMoneyException {
+        if(sign.getLine(3).startsWith(tl("signFormatWorth")) && updateFromWorth(sign, player, ess, ess.getSettings().getEcoBuyMultiplier())) {
```
```suggestion
        if (sign.getLine(3).startsWith(tl("signFormatWorth")) && updateFromWorth(sign, player, ess, ess.getSettings().getEcoBuyMultiplier())) {
```

mdcfe commented on Essentials/src/com/earth2me/essentials/signs/SignSell.java:
```diff
@@ -18,13 +20,21 @@ public SignSell() {
-        validateTrade(sign, 1, 2, player, ess);
-        validateTrade(sign, 3, ess);
+        validateTrade(sign, 1, 2, player, ess);  
+        if(sign.getLine(3).isEmpty()) {
+            updateFromWorth(sign, player, ess, new BigDecimal("1.0"));
+        }
+        else {
```
```suggestion
        if (sign.getLine(3).isEmpty()) {
            updateFromWorth(sign, player, ess, new BigDecimal("1.0"));
        } else {
```

This is also common functionality between buy and sell which ought to be in a separate class.

mdcfe commented on Essentials/src/com/earth2me/essentials/signs/SignSell.java:
```diff
@@ -18,13 +20,21 @@ public SignSell() {
 
     @Override
     protected boolean onSignInteract(final ISign sign, final User player, final String username, final IEssentials ess) throws SignException, ChargeException, MaxMoneyException {
+        if(sign.getLine(3).startsWith(tl("signFormatWorth")) && updateFromWorth(sign, player, ess, new BigDecimal("1.0"))) {
+            return false;
+        }
```
```suggestion
        if (sign.getLine(3).startsWith(tl("signFormatWorth")) && updateFromWorth(sign, player, ess, new BigDecimal("1.0"))) {
            return false;
        }
```

Ditto (also, use an overloaded method for the default scaling factor).

mdcfe commented on Essentials/src/messages.properties:
```diff
@@ -592,6 +593,7 @@ ptimeCommandDescription=Adjust player's client time. Add @ prefix to fix.
 ptimeCommandUsage=/<command> [list|reset|day|night|dawn|17:30|4pm|4000ticks] [player|*]
 pweatherCommandDescription=Adjust a player's weather
 pweatherCommandUsage=/<command> [list|reset|storm|sun|clear] [player|*]
+priceChanged=\u00a76The price of \u00a7c{0}x {1}\u00a76 has changed from \u00a7e{2}\u00a76 to \u00a7e{3}\u00a76.
```
We should prompt the player to right-click the sign again if they want to buy/sell at the new price.

moxventura replied:
Done

mdcfe commented on Essentials/src/messages.properties:
```diff
@@ -584,6 +584,7 @@ powerToolRemove=\u00a76Command \u00a7c{0}\u00a76 removed from \u00a7c{1}\u00a76.
 powerToolRemoveAll=\u00a76All commands removed from \u00a7c{0}\u00a76.
 powerToolsDisabled=\u00a76All of your power tools have been disabled.
 powerToolsEnabled=\u00a76All of your power tools have been enabled.
+priceChanged=\u00a76The price of \u00a7c{0}x {1}\u00a76 has changed from \u00a7c{2}\u00a76 to \u00a7c{3}\u00a76.
```
```suggestion
```

Duplicate message

moxventura replied:
Must have happened with a merge conflict 👎

Commit:
Resolve md678685s comments

Thx!

moxventura commented:
I'm willing to update this again, but only if the choice is made to implement this feature.

JRoy commented:
Pretty sure the consensus on this PR is that in order to do this perfectly, Essentials would need to store the locations of all signs and that is something we don't want to do. This solution of updating the signs on click is also not a good solution as it would be misleading to users. Thanks for your interest in contributing but we will reject this PR at this time

Techno3600 commented:
With all due respect, I completely disagree with this decision. This option is editable in the configuration, and thus people who really need this for their server could enable it, knowing that it will not update immediately. The alternative was hours upon hours of work for my team manually updating signs every time we made a change in worth.yml back when we had that season running. Our team, and I'm sure a lot of others, would easily take the option of signs updating onclick over hours of manually editing sings. Also, the multiplayer addition in this PR would have made our team's life so much easier because we wouldn't have had to pull out a calculator literally every single time we wanted to update the price.

I would strongly urge @JRoy and the other creators of EssentialsX to reconsider this decision and possibly add this into the master project.
